closed/src/java.base/share/classes/com/ibm/security/criu/SHA.java

```diff
@@ -0,0 +1,239 @@
+ *
+ * <p>It implement JavaSecurity MessageDigest, and can be used by in
+ * the Java Security framework, as a pluggable implementation, as a
+ * filter for the digest stream classes.
+ *
+ * @author      Roger Riggs
+ * @author      Benjamin Renaud
+ * @author      Andreas Sterbenz
+ */
+public final class SHA extends DigestBase {
+
+    // Buffer of int's and count of characters accumulated
+    // 64 bytes are included in each hash block so the low order
+    // bits of count are used to know how to pack the bytes into ints
+    // and to know when to compute the block and start the next one.
+    private int[] W;
+
+    // state of this
+    private int[] state;
+
+    /**
+     * Creates a new SHA object.
+     */
+    public SHA() {
+        super("SHA-1", 20, 64);
+        state = new int[5];
+        W = new int[80];
+        resetHashes();
+    }
+
+    /*
+     * Clones this object.
+     */
+    public Object clone() throws CloneNotSupportedException {
+        SHA copy = (SHA) super.clone();
+        copy.state = copy.state.clone();
+        copy.W = new int[80];
+        return copy;
+    }
+
+    /**
+     * Resets the buffers and hash value to start a new hash.
+     */
+    void implReset() {
+        // Load magic initialization constants.
+        resetHashes();
+        // clear out old data
+        Arrays.fill(W, 0);
+    }
+
+    private void resetHashes() {
+        state[0] = 0x67452301;
+        state[1] = 0xefcdab89;
+        state[2] = 0x98badcfe;
+        state[3] = 0x10325476;
+        state[4] = 0xc3d2e1f0;
+    }
+
+    /**
+     * Computes the final hash and copies the 20 bytes to the output array.
+     */
+    void implDigest(byte[] out, int ofs) {
+        long bitsProcessed = bytesProcessed << 3;
+
+        int index = (int) bytesProcessed & 0x3f;
+        int padLen = (index < 56) ? (56 - index) : (120 - index);
+        engineUpdate(padding, 0, padLen);
+
+        try {
+            // Invoke the i2bBig4 method from the ByteArrayAccess class.
+            Class<?> runnable = Class.forName("sun.security.provider.ByteArrayAccess",
+                    true, ClassLoader.getSystemClassLoader());
+            Method i2bBig4 = runnable.getDeclaredMethod("i2bBig4",
+                    int.class, byte[].class, int.class);
+            i2bBig4.setAccessible(true);
+            i2bBig4.invoke(null, (int) (bitsProcessed >>> 32), buffer, 56);
+            i2bBig4.invoke(null, (int) bitsProcessed, buffer, 60);
+            implCompress(buffer, 0);
+
+            // Invoke the i2bBig method from the ByteArrayAccess class.
+            Method i2bBig = runnable.getDeclaredMethod("i2bBig",
+                    int[].class, int.class, byte[].class, int.class, int.class);
+            i2bBig.setAccessible(true);
+            i2bBig.invoke(null, state, 0, out, ofs, 20);
+        } catch (Exception e) {
+            System.out.println(e.toString());
+        }
+    }
+
+    // Constants for each round
```

Comment:
This is probably better as a sentence fragment (without a capital at the beginning).